datanode/src/main/java/com/odysseusinc/arachne/datanode/util/ZipUtils.java

```diff
@@ -0,0 +1,130 @@
+import static java.nio.file.FileVisitResult.CONTINUE;
+import static java.nio.file.StandardOpenOption.CREATE;
+import static java.nio.file.StandardOpenOption.WRITE;
+
+public class ZipUtils {
+    public static byte[] zipFile(String name, byte[] content) {
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        try (ZipOutputStream zip = new ZipOutputStream(out)) {
+            addZipEntry(zip, name, content);
+        } catch (IOException e) {
+            throw new RuntimeException("Error packing file [" + name + "] of " + content.length + " bytes: " +  e);
+        }
+        return out.toByteArray();
+    }
+
+    public static void zip(Path archiveFile, Path path) throws IOException {
+        try (
+                OutputStream out = Files.newOutputStream(archiveFile, CREATE, WRITE);
+                ZipOutputStream zip = new ZipOutputStream(out)
+        ) {
+            if (Files.isDirectory(path)) {
+                Files.walkFileTree(path, new AddZipEntryFileVisitor(path, zip));
+            } else {
+                addZipEntry(zip, path, path.getFileName().toString());
+            }
+        }
+    }
+
+    @SafeVarargs
+    public static Map<Function<String, FailableFunction<InputStream, Object, IOException>>, Map<String, Object>> processZip(
+            InputStream source,
+            Function<String, FailableFunction<InputStream, Object, IOException>>... converters
+    ) {
+        try (ZipArchiveInputStream zip = new ZipArchiveInputStream(source, UTF_8.name(), true, false, true)) {
+            Map<Function<String, FailableFunction<InputStream, Object, IOException>>, Map<String, Object>> results = new HashMap<>();
+            ArchiveEntry entry;
+            while ((entry = zip.getNextEntry()) != null) {
+                if (!entry.isDirectory()) {
+                    String name = entry.getName();
+                    for (Function<String, FailableFunction<InputStream, Object, IOException>> converter : converters) {
+                        FailableFunction<InputStream, Object, IOException> found = converter.apply(name);
+                        if (found != null) {
+                            results.computeIfAbsent(converter, __ -> new HashMap<>()).put(name, found.apply(new NoCloseInputStream(zip)));
+                            break;
+                        }
+                    }
+                }
+            }
+            return results;
+        } catch (IOException e) {
+            throw new RuntimeException("Error processing zip", e);
+        }
+    }
+
+    private static class NoCloseInputStream extends ProxyInputStream {
+        public NoCloseInputStream(InputStream proxy) {
+            super(proxy);
+        }
+
+        @Override
+        public void close() throws IOException {
+        }
+    }
+
+    private static void addZipEntry(ZipOutputStream zos, Path file, String realName) throws IOException {
+        addZipEntry(zos, realName, Files.readAllBytes(file));
```

Comment:
That can be OutOfMemoryError for big files.

Reply:
Adjusted